Makefile

```diff
@@ -30,3 +30,89 @@ binaries:
 
 	mkdir -p bin/arm64
 	GOOS=darwin GOARCH=arm64 go build -o bin/arm64/idemixgen tools/idemixgen/main.go
+
+.PHONY: deps
+deps: $(BUF) $(PROTOC_GEN_GO)
+
+PROJECT := idemix
+
+# This controls the version of buf to install and use.
+BUF_VERSION := 1.1.1
+PROTOC_VERSION := 3.19.4
+PROTOC_GEN_GO_VERSION := v1.3.2
+
+UNAME_OS := $(shell uname -s)
+UNAME_ARCH := $(shell uname -m)
+ifeq ($(UNAME_OS),Darwin)
+	PLATFORM := osx
+endif
+ifeq ($(UNAME_OS),Linux)
+	PLATFORM := linux
+endif
+
+# Buf will be cached to ~/.cache/buf-example.
+CACHE_BASE := $(HOME)/.cache/$(PROJECT)
+# This allows switching between i.e a Docker container and your local setup without overwriting.
+CACHE := $(CACHE_BASE)/$(UNAME_OS)/$(UNAME_ARCH)
+# The location where buf will be installed.
+CACHE_BIN := $(CACHE)/bin
+# Marker files are put into this directory to denote the current version of binaries that are installed.
+CACHE_VERSIONS := $(CACHE)/versions
+
+# Update the $PATH so we can use buf directly
+export PATH := $(abspath $(CACHE_BIN)):$(PATH)
+
+
+# BUF points to the marker file for the installed version.
+#
+# If BUF_VERSION is changed, the binary will be re-downloaded.
+BUF := $(CACHE_VERSIONS)/buf/$(BUF_VERSION)
+$(BUF):
+	@rm -f $(CACHE_BIN)/buf
+	@mkdir -p $(CACHE_BIN)
+	curl -sSL \
+		"https://github.com/bufbuild/buf/releases/download/v$(BUF_VERSION)/buf-$(UNAME_OS)-$(UNAME_ARCH)" \
+		-o "$(CACHE_BIN)/buf"
+	chmod +x "$(CACHE_BIN)/buf"
+
+	@rm -rf $(dir $(BUF))
+	@mkdir -p $(dir $(BUF))
+	@touch $(BUF)
+# PROTOC_GEN_GO points to the marker file for the installed version.
+#
+# If PROTOC_GEN_GO_VERSION is changed, the binary will be re-downloaded.
+PROTOC_GEN_GO := $(CACHE_VERSIONS)/protoc-gen-go/$(PROTOC_GEN_GO_VERSION)
+$(PROTOC_GEN_GO):
+	@rm -f $(CACHE_BIN)/protoc-gen-go
+	@mkdir -p $(CACHE_BIN)
+	$(eval PROTOC_GEN_GO_TMP := $(shell mktemp -d))
+	cd $(PROTOC_GEN_GO_TMP); go install github.com/golang/protobuf/protoc-gen-go@$(PROTOC_GEN_GO_VERSION)
+	@rm -rf $(PROTOC_GEN_GO_TMP)
+	@rm -rf $(dir $(PROTOC_GEN_GO))
+	@mkdir -p $(dir $(PROTOC_GEN_GO))
+	@touch $(PROTOC_GEN_GO)	
+
+# PROTOC points to the marker file for the installed version.
+#
+# If PROTOC_VERSION is changed, the binary will be re-downloaded.
+PROTOC := $(CACHE_VERSIONS)/protoc/$(PROTOC_VERSION)
+$(PROTOC):
+	@rm -f $(CACHE_BIN)/protoc
+	@mkdir -p $(CACHE_BIN)
+	$(eval PROTOC_TMP := $(shell mktemp -d))
+	curl -sSL \
+		"https://github.com/protocolbuffers/protobuf/releases/download/v$(PROTOC_VERSION)/protoc-$(PROTOC_VERSION)-$(PLATFORM)-$(UNAME_ARCH).zip" \
+		-o "$(PROTOC_TMP)/protoc.zip"
+	unzip -o "$(PROTOC_TMP)/protoc.zip" -d "$(CACHE)" bin/protoc
+	unzip -o "$(PROTOC_TMP)/protoc.zip" -d "$(CACHE)" include/*
+	@rm -rf $(PROTOC_TMP)
+	chmod +x "$(CACHE_BIN)/protoc"
+	@rm -rf $(dir $(PROTOC))
+	@mkdir -p $(dir $(PROTOC))
+	@touch $(PROTOC)
```

Comment:
I'm guessing that you already have all the dependencies installed. A fair bit of the Makefile (which is based on a sample from Buf) is to ensure that the builds are repeatable, with defined versions in a specific location rather than picking up system versions etc.

Reply:
no, I removed everything in the `pkg` and `bin` directories before trying. And it also should work, because go install will build a binary, go binaries are static binaries and so they must build against the necessary dependencies. So we can go get all sorts of other tools/dependencies at versions we like, but when we go build buf at a specific version, behind the curtains the compiler will pull all the necessary dependencies for that build and run it.

To sum up, I still see no reason to have anything else in the makefile other than those two lines (but of course I'm happy to be proven wrong).